storage/transfermanager/downloader.go

```diff
@@ -306,3 +396,76 @@ type DownloadOutput struct {
 	Err    error                      // error occurring during download
 	Attrs  *storage.ReaderObjectAttrs // attributes of downloaded object, if successful
 }
+
+// TODO: use built-in after go < 1.21 is dropped.
+func min(a, b int64) int64 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// numShards calculates how many shards the given range should be divided into
+// given the part size.
+func numShards(attrs *storage.ReaderObjectAttrs, r *DownloadRange, partSize int64) int {
+	objectSize := attrs.Size
+
+	// Transcoded objects do not support ranged reads.
+	if attrs.ContentEncoding == "gzip" {
+		return 1
+	}
+
+	if r == nil {
+		// Divide entire object into shards.
+		return int(math.Ceil(float64(objectSize) / float64(partSize)))
+	}
+	// Negative offset reads the whole object in one go.
+	if r.Offset < 0 {
+		return 1
+	}
+
+	firstByte := r.Offset
+
+	// Read to the end of the object, or, if smaller, read only the amount
+	// of bytes requested.
+	lastByte := min(objectSize-1, r.Offset+r.Length-1)
```

Comment:
Is this really relevant if we are just calculating number of shards? The math seems a bit overcomplicated.

Reply:
It's very possible it's overcomplicated, I'll take a look and see if I can make it simpler.

Reply:
So I can maybe make the code easier to follow (probably by removing that use of `min` and replacing with a conditional) and pop it out into a separate function, but given the edge cases I don't see a less roundabout way of calculating the number of shards. Am I missing something?

Reply:
I think this is probably fine, not sure what I was thinking of earlier...